File: xcode/tests/ptrArrayTest.m
/*
 *	Generated 07/02/2015 10:57:54
 */


/*
 This is free and unencumbered software released into the public domain.
 
 Anyone is free to copy, modify, publish, use, compile, sell, or
 distribute this software, either in source code form or as a compiled
 binary, for any purpose, commercial or non-commercial, and by any
 means.
 
 In jurisdictions that recognize copyright laws, the author or authors
 of this software dedicate any and all copyright interest in the
 software to the public domain. We make this dedication for the benefit
 of the public at large and to the detriment of our heirs and
 successors. We intend this dedication to be an overt act of
 relinquishment in perpetuity of all present and future rights to this
 software under copyright law.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 OTHER DEALINGS IN THE SOFTWARE.
 
 For more information, please refer to <http://unlicense.org/>
 */


#import <XCTest/XCTest.h>
#import <libCmn/libCmn.h>



// All code under test must be linked into the Unit Test bundle
// Test Macros:
//      XCTAssert(expression, failure_description, ...)
//      XCTFail(failure_description, ...)
//      XCTAssertEqualObjects(object_1, object_2, failure_description, ...) uses isEqualTo:
//      XCTAssertEquals(value_1, value_2, failure_description, ...)
//      XCTAssertEqualsWithAccuracy(value_1, value_2, accuracy, failure_description, ...)
//      XCTAssertNil(expression, failure_description, ...)
//      XCTAssertNotNil(expression, failure_description, ...)
//      XCTAssertTrue(expression, failure_description, ...)
//      XCTAssertFalse(expression, failure_description, ...)
//      XCTAssertThrows(expression, failure_description, ...)
//      XCTAssertThrowsSpecific(expression, exception_class, failure_description, ...)
//      XCTAssertThrowsSpecificNamed(expression, exception_class, exception_name, failure_description, ...)
//      XCTAssertNoThrow(expression, failure_description, ...)
//      XCTAssertNoThrowSpecific(expression, exception_class, failure_description, ...)
//      XCTAssertNoThrowSpecificNamed(expression, exception_class, exception_name, failure_description, ...)


#include    "ptrArray_internal.h"

static
char	*stringTable[] = {
    "now",
    "before",
    "after",
    "tomorrow",
    "today",
    "someday",
};
static
int             num = 6;


@interface ptrArrayTests : XCTestCase

@end

@implementation ptrArrayTests


- (void)setUp
{
    [super setUp];
    // Put setup code here. This method is called before the invocation of each
    // test method in the class.
    mem_Init();
    
    
}


- (void)tearDown
{
    mem_Dump( );
    mem_Release( );
    
    // Put teardown code here. This method is called after the invocation of each
    // test method in the class.
    [super tearDown];
}




- (void)testOpenClose
{
    PTRARRAY_DATA	*pObj = OBJ_NIL;
    ERESULT         eRc;
    uint32_t        i;
    char            *pStr;
   
    pObj = ptrArray_Alloc(0);
    XCTAssertFalse( (OBJ_NIL == pObj), @"" );
    pObj = ptrArray_Init( pObj );
    XCTAssertFalse( (OBJ_NIL == pObj) );
    if (pObj) {
        
        for (i=0; i<num; ++i) {
            eRc = ptrArray_AppendData(pObj, stringTable[i], NULL);
            XCTAssertTrue( (ERESULT_IS_SUCCESSFUL(eRc)) );
        }
        
        for (i=0; i<num; ++i) {
            pStr = ptrArray_GetData(pObj, i+1);
            XCTAssertFalse( (NULL == pStr) );
            XCTAssertTrue( (0 == strcmp(pStr, stringTable[i])) );
        }
        
        eRc = ptrArray_Sort(pObj, (PTR_COMPARE)&strcmp);
        
        for (i=0; i<num; ++i) {
            pStr = ptrArray_GetData(pObj, i+1);
            fprintf( stderr, "%d - %s\n", i, pStr );
        }
        
        obj_Release(pObj);
        pObj = OBJ_NIL;
    }

}



@end

File: tests/ptrArray_test.c
/*
 *	Generated 06/05/2017 21:57:10
 */





// All code under test must be linked into the Unit Test bundle
// Test Macros:
//      TINYTEST_ASSERT(condition)
//      TINYTEST_ASSERT_MSG(condition,msg)
//      TINYTEST_EQUAL(expected, actual)
//      TINYTEST_EQUAL_MSG(expected, actual, msg)
//      TINYTEST_FALSE_MSG(condition,msg)
//      TINYTEST_FALSE(condition)
//      TINYTEST_TRUE_MSG(pointer,msg)
//      TINYTEST_TRUE(condition)





#include    <tinytest.h>
#include    <cmn_defs.h>
#include    <trace.h>
#include    <ptrArray_internal.h>



static
char	*stringTable[] = {
    "now",
    "before",
    "after",
    "tomorrow",
    "today",
    "someday",
};
static
int             num = 6;




int         setUp(
    const
    char        *pTestName
)
{
    mem_Init( );
    trace_Shared( ); 
    // Put setup code here. This method is called before the invocation of each
    // test method in the class.
    
    return 1; 
}


int         tearDown(
    const
    char        *pTestName
)
{
    // Put teardown code here. This method is called after the invocation of each
    // test method in the class.

    
    trace_SharedReset( ); 
    if (mem_Dump( ) ) {
        fprintf(
                stderr,
                "\x1b[1m"
                "\x1b[31m"
                "ERROR: "
                "\x1b[0m"
                "Leaked memory areas were found!\n"
        );
        exitCode = 4;
        return 0;
    }
    mem_Release( );
    
    return 1; 
}




int    	compareStrings(
    const
    char        *pStr1,
    const
    char        *pStr2
)
{
    int         cmp = strcmp(pStr1, pStr2);
    return cmp;
}






int         test_ptrArray_OpenClose(
    const
    char            *pTestName
)
{
    PTRARRAY_DATA   *pObj = OBJ_NIL;
    ERESULT         eRc;
    uint32_t        i;
    uint32_t        index;
    const
    char            *pStr;
   
    fprintf(stderr, "Performing: %s\n", pTestName);
    pObj = ptrArray_Alloc( );
    TINYTEST_FALSE( (OBJ_NIL == pObj) );
    pObj = ptrArray_Init(pObj);
    TINYTEST_FALSE( (OBJ_NIL == pObj) );
    if (pObj) {

        for (i=0; i<num; ++i) {
            eRc = ptrArray_AppendData(pObj, stringTable[i], &index);
            XCTAssertTrue( (ERESULT_IS_SUCCESSFUL(eRc)) );
            XCTAssertTrue( (index == (i + 1)) );
        }
        
        for (i=0; i<ptrArray_getSize(pObj); ++i) {
            pStr = ptrArray_GetData(pObj, i+1);
            XCTAssertFalse( (NULL == pStr) );
            XCTAssertTrue( (0 == strcmp(pStr, stringTable[i])) );
        }
        
        eRc = ptrArray_Sort(pObj, (PTR_COMPARE)&compareStrings);
        
        fprintf(stderr, "\n\nSorted Array:\n");
        for (i=0; i<ptrArray_getSize(pObj); ++i) {
            pStr = ptrArray_GetData(pObj, i+1);
            fprintf(stderr, "%d - %s\n", i+1, pStr);
        }
        fprintf(stderr, "End of Sorted Array\n\n");
        
        pStr = (const char *)ptrArray_Delete(pObj, 2);
        fprintf(stderr, "Deleted: '%s'\nArray After Delete:\n", pStr);
        for (i=0; i<ptrArray_getSize(pObj); ++i) {
            pStr = ptrArray_GetData(pObj, i+1);
            fprintf(stderr, "%d - %s\n", i+1, pStr);
        }
        fprintf(stderr, "End of Array After Delete\n\n");
        
        obj_Release(pObj);
        pObj = OBJ_NIL;
    }

    fprintf(stderr, "...%s completed.\n", pTestName);
    return 1;
}




TINYTEST_START_SUITE(test_ptrArray);
  TINYTEST_ADD_TEST(test_ptrArray_OpenClose,setUp,tearDown);
TINYTEST_END_SUITE();

TINYTEST_MAIN_SINGLE_SUITE(test_ptrArray);





